### PR TITLE
Add many wooden fences made out of the 4 kinds of woods

### PR DESCRIPTION
**Pull request by Lymkwi**
- Add fences made out of the four kinds of wood registered in default, with their craft recipe

Wooden Fences

### CODE REVIEW

**Comment by paramat** [comment]
+1.

**Comment by kilbith** [comment]
C'mon, complete that track on : boat, doors, shelves, rail/ladder...

:-1:

**Comment by PilzAdam** [comment]
:-1:, this is a mod thing.

**Comment by LeMagnesium** [comment]
I usually agree when people comment on Minetest Game's pull requests that "nha, not that, it could be done by a mod , -1", but about fences, I consider that default already registers a wooden fence, and has a lot of other kinds of woods, that are not pretty useful. By that I mean that you could easily make all of vanilla's wooden nodes using only `default:wood`, and never use the other woods. If acacias, pines and jungle trees are in vanilla, they should be used to make different sort of wooden nodes like fences, maybe ladders, otherwise they're only registered to be placed for trees, or to be used as nodes belonging to `group:wood`, and we could easily remove them, their trees, and replace all `group:wood` with `default:wood`. I don't need to remind everyone that the first thing people say after trying vanillla is that it lacks nodes, forcing us to use mods, and it always brings the usual debate about "should it be in a mod, or should it be in vanilla"? (Then again if vanilla is a "base" for mods then move back acacias and pines into `moretrees` mod).
To sum it up: If we have different woods, we should be able to do specific things with each one of them, otherwise remove the other woods and you do everything with only trees.
However, I understand why you can consider that it's a mod thing, there is already a mod dedicated to making different kind of fences, a quite old one.
In the end I just want to say that I'm not particulary defending that feature, it's just something that was requested, if you don't want it then go to its issue (#663), say it there, close it, and I will automatically remove that pull request.

**Comment by Cyberpangolin** [comment]
:+1:  
Please include it! I'm no programmer, I'm not a leader in the minetest project, and I don't intend to be. I'm only a player. 
But as a player, I think we need a real game once, without any mod. Such fences are really simple and won't change the game-play. Why should I always take mods for every single feature, with the risk of incompatibility?

**Comment by rubenwardy** [comment]
Don't stairs support all the woods?

**Comment by RealBadAngel** [comment]
Agree, adding new content without use for it is pointless. mt game screams out loud for no content at all - and dont say here its "mod stuff". Its just a bad joke :P

**Comment by paramat** [comment]
There was a PR for new fences #297 so sfan5 might support this.
I perhaps wouldn't go as far as doors, boats, bookcases, ladders etc. Fences though seem an obvious one to use more woods, they also look really good whereas just the one standard wood can get boring.

**Comment by kilbith** [comment]
> I perhaps wouldn't go as far as doors, boats, bookcases, ladders etc. Fences though seem an obvious one to use more woods

That's an arbitrarily drawn line.

**Comment by LeMagnesium** [comment]
You can find a limit for that line then. I agree that doors, boats, bookcases, ladders, sticks, etc, made out of all 4 woods would be too much nodes. *However*, creating variations of ladders, or sticks, using the 4 woods would be enough. If you want a logical limit, then let's restrict those variations to default. Does it seem okay to you?
For now, our only issue here are wooden fences, we can see what else could be added (or not) later.

**Comment by paramat** [comment]
If the function was made a global function and a groups field added this would allow users to register fences in any material. It would work like 'register stairs' in the stairs mod. This would also be closer to sfan5's original PR #297 so perhaps more chance of it being accepted.

function default.register_fence(name, desc, groups)

**Comment by LeMagnesium** [comment]
Good idea, I didn't think it would be used outside of default but why not.
[Done](https://github.com/LeMagnesium/minetest_game/commit/b66bd93942d0649015dab904fd3e796467d5d378)

**Comment by nerzhul** [comment]
:+1: with this PR because wood is in default, no reason to have a separate mod, it's a game, not minimal.

**Comment by C1ffisme** [comment]
I think it looks okay. It's not really going that far.

Consider this, would you like your junglewood house to be contrasted by a normal colored fence around it?

:+1:

**Comment by Ferk** [comment]
An arbitrary line was already drawn the moment different types of wood craft into different types of wooden planks.

The textures are there already. Imho, there's not much to lose when using it for different fences. I guess there's no reason why an item that doesn't require an entirely new texture couldn't use the texture from the wooden plank it comes from.

If the item requires a new texture, it could have a neutral color and consider it sort of a wood conglomerate.

**Comment by Sokomine** [comment]
More fences are welcome. The diffrent wood types minetest_game got are a huge improvement and contribute to players creating more intresting buildings. Fences that fit to the house (or, in some situations, produce a nice contrast) offer far more choice for creations. They may also be used as indigrents for craft receipes. And let's not forget that fence posts can creatively be used for other purposes than fences.
I'm working on a trees_lib mod that will hopefully fascilitate adding diffrent trees through mods and also create diffrently shaped wooden nodes if desired.

**Comment by VanessaE** [comment]
I favor adding fences for each type of tree for the sake of variety.

This isn't the first time someone's wanted to add a few nodes to the game, and it won't be the last, so if someone thinks this sort of feature belongs in a mod, that's fine, but that means there needs to be a clear, unambiguous line somewhere that mt_game will not cross when the next "let's add X" comes up.

**Comment by kilbith** [comment]
-1 of _sfan5_ : http://irc.minetest.ru/minetest-dev/2015-09-20#i_4400232

**Comment by LeMagnesium** [comment]
I just updated the commit to include a `fence` group that would allow mods to recognize easily a `group:fence` (just like `group:stick`).

**Comment by paramat** [comment]
Unfortunately this is not popular with the MTGame devs at the moment.

**Comment by C1ffisme** [comment]
This could be an instant merge if we could just set the texture depending on some node variable.

**Comment by LeMagnesium** [comment]
The texture used is determined using the first parameter given to the function ([that line](https://github.com/minetest/minetest_game/pull/665/files#diff-78f72d86b6ccbcf349905228f797475eR24)). If you don't like this method I could add another argument which would be a specific texture name.

**Comment by Ferk** [comment]
I doubt the way it's implemented has anything to do with the popularity of this PR among the devs.

I believe they think this would be better as a mod.

**Comment by BlockMen** [comment]
Added in Minetest NeXt.
https://github.com/BlockMen/minetest_next/commit/b25477b3151bce34dfa66607843c14862cabbf90

**Comment by LeMagnesium** [comment]
While looking around the issue that this pull request is supposed to 'fix', I found out that the request for more fences (one for each wood) originated from a post on the forums; with a poll asking if people agreed to add more fences.

> Unfortunately this is not popular with the MTGame devs at the moment.

In fact, it may not be popular among the developers (yet not all), but players want it : [see the poll's results](https://forum.minetest.net/viewtopic.php?f=5&t=13162&view=viewpoll)
If you decide that you don't want it though, it's fine. Tell it clearly, and I will close this immediatly.

**Comment by ShadowNinja** [comment]
I like the idea, but I don't like the weird connections between different types and I think the craft has to give more than two fences now that it requires more than four full nodes worth of material.

**Comment by paramat** [comment]
So far between the 5 mtgame devs: 2 for, 2 against.

**Comment by kilbith** [comment]
> So far between the 5 mtgame devs: 2 for, 2 against.

You really should stop cheating on the votes... Agreeing on an idea != agreeing on an implementation.

Btw this PR needs the label [controversial], tacitly agreed these PRs never pass the merging step.

**Comment by LeMagnesium** [comment]
> I like the idea, but I don't like the weird connections between different types

I agree that it looks odd. I don't know any way to tell the engine not to connect the fences together except removing the `fencelike` drawtype and handling in the mod itself the connections using different nodes with different nodebox shapes (like in the fences mod), but that would be imprecise, heavy, and add tons of nodes. I think a system similar to the`connect_to_raillike` group would be helpful in order to do such a thing.

> Btw this PR needs the label [controversial], tacitly these PRs never pass the merging step.

Even though the [original PR](https://github.com/minetest/minetest_game/pull/297) was labeled as controversial, adding the controversial label to this one means to me that you're most likely not going to merge it. I would rather see it closed than labeled as [controversial](https://github.com/minetest/minetest_game/pulls?utf8=%E2%9C%93&q=is%3Apr+label%3Acontroversial).

**Comment by paramat** [comment]
> Agreeing on an idea != agreeing on an implementation.

Yes fair enough, 2 like the idea and 2 do not.
Before i just forgot about sfan5's vote, i wasn't cheating, calm down.

Devs: perhaps this should be closed as it will probably not get anywhere?

**Comment by kwolekr** [comment]
I can't help but agree with MG's rationale here.  Adding more content is useless if you can't do anything with it.
It also highlights the need for a mechanism to separate node types from node materials.  We had a brief discussion about it on IRC but it never got past the idea phase.

This can always be removed in the future, no?  I don't see any actual problem with just adding them for now, and in the future when you could define a set of possible materials for a certain type of node, it can quite trivially be removed.

**Comment by Ekdohibs** [comment]
I think this could be merged, and maybe removed later if we have a better way to do it (like materials), so :+1:

**Comment by sofar** [comment]
Please merge this. The fact that there are several mods out there that implement this is a clear indication that this is both *needed* and *wanted*.

**Comment by paramat** [comment]
Seems to be 3 for 2 against, it's a majority but is 2 against enough to block something? I dunno =/. Personally i think this should be merged it seems to have a lot of non-MTgame-dev support.
A 'register fence' function allows fences in any material, like stairs, seems useful.
> but I don't like the weird connections between different types and I think the craft has to give more than two fences now that it requires more than four full nodes worth of material.

Agreed, differing colours of fences should not connect, if that's unavoidable then probably not a problem as differing colours would usually be kept well apart.

**Comment by VanessaE** [comment]
It seems to me that the different-colors solution is simple:

Instead of extending fence rungs out from the center of one fence to the center of its neighbor a meter away, extend them half a meter out from *both* adjoining nodes, i.e. so that the two parts meet at the +/- 0.5 node edge.  

Add some shape and texture detail where the two parts meet, so that it looks like tongue-and-groove, dovetail, or another "proper" joint.

**Comment by sofar** [comment]
#743 - alternative implementation I wrote

Not to dig up a dead horse, or beat it. Here is a summary of why I think this is still worth it:

- #743 retains original fence recipe and just adds new sticks. This makes a lot more sense gameplay wise, is easier to remember, and fills up bags which is good for gameplay (forces player to manage bags)
- the API is a bit more robust IMO, and I've properly documented it

As for the weird connection - this can and should be fixed in content_mapblock.cpp in minetest, so that should not preclude merging this patch.

Drawing fence nodeboxes in a mod sounds horribly wasteful to me, core handles this efficiently already, so this should stay in core (I'd like to add NDT_WALLLIKE as well myself).

**Comment by sofar** [comment]
BTW I've seen custom fences like this on at least 10 or so publicly listed servers. A good argument that there is a use for it. Some standardization with a decent API will help a lot of people.

**Comment by paramat** [comment]
Yeah #743 is better, no weird mixing of applewood sticks and non-applewood wood nodes.

**Comment by LeMagnesium** [comment]
I close this, please refer to #743 , which is in my opinion a better implementation of fences.
